driver/src/main/com/mongodb/client/gridfs/GridFSUploadStreamImpl.java

```diff
@@ -0,0 +1,165 @@
+import static com.mongodb.util.Util.toHex;
+
+final class GridFSUploadStreamImpl extends GridFSUploadStream {
+    private final MongoCollection<BsonDocument> filesCollection;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private final ObjectId fileId;
+    private final String filename;
+    private final int chunkSizeBytes;
+    private final BsonDocument metadata;
+    private final MessageDigest md5;
+    private boolean closed;
+    private byte[] buffer;
+    private long lengthInBytes;
+    private int bufferOffset;
+    private int chunkIndex;
+
+    GridFSUploadStreamImpl(final MongoCollection<BsonDocument> filesCollection, final MongoCollection<BsonDocument> chunksCollection,
+                           final ObjectId fileId, final String filename, final int chunkSizeBytes, final BsonDocument metadata) {
+        this.filesCollection = notNull("files collection", filesCollection);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+        this.fileId = notNull("File Id", fileId);
+        this.filename = notNull("filename", filename);
+        this.chunkSizeBytes = chunkSizeBytes;
+        this.metadata = metadata;
+        try {
+            md5 = MessageDigest.getInstance("MD5");
+        } catch (NoSuchAlgorithmException e) {
+            throw new MongoGridFSException("No MD5 message digest available, cannot upload file", e);
+        }
+        chunkIndex = 0;
+        bufferOffset = 0;
+        buffer = new byte[chunkSizeBytes];
+    }
+
+    @Override
+    public ObjectId getFileId() {
+        return fileId;
+    }
+
+    @Override
+    public void write(final int b) {
+        byte[] byteArray = new byte[1];
+        byteArray[0] = (byte) (0xFF & b);
+        write(byteArray, 0, 1);
+    }
+
+    @Override
+    public void write(final byte[] b) {
+        write(b, 0, b.length);
+    }
+
+    @Override
+    public void write(final byte[] b, final int off, final int len) {
+        checkClosed();
+        if (b == null) {
+            throw new NullPointerException();
+        } else if ((off < 0) || (off > b.length) || (len < 0)
+                || ((off + len) > b.length) || ((off + len) < 0)) {
+            throw new IndexOutOfBoundsException();
+        } else if (len == 0) {
+            return;
+        }
+
+        int currentOffset = off;
+        int lengthToWrite = len;
+        int amountToCopy = 0;
+
+        while (lengthToWrite > 0) {
+            amountToCopy = lengthToWrite;
+            if (amountToCopy > chunkSizeBytes - bufferOffset) {
+                amountToCopy = chunkSizeBytes - bufferOffset;
+            }
+            System.arraycopy(b, currentOffset, buffer, bufferOffset, amountToCopy);
+
+            bufferOffset += amountToCopy;
+            currentOffset += amountToCopy;
+            lengthToWrite -= amountToCopy;
+            lengthInBytes += amountToCopy;
+
+            if (bufferOffset == chunkSizeBytes) {
+                writeChunk();
```

Comment:
Do you want to do any batching of chunk writing?

Looks like this is writing them one at a time.